first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1235 @@
+        return
+    for y in x.after:
+        follow(following_layers, layer_graph, layer_graph[y])
+
+
+def get_first_conv(layer_graph, following_layers):
+    """
+    Gets first convolution in following layers
+    """
+    for layer in following_layers:
+        if layer_graph[layer].type == 'conv' and layer_graph[layer].value.groups == 1:
+            return layer
+
+def compute_deletion(layer_graph,
+                     custom_deletions,
+                     deleted_things,
+                     layer,
+                     sort,
+                     custom,
+                     reason=None):
+    """
+    Given a layer, generate the list of the indexes we need to delete from its following layers
+    """
+    # find all the following layers
+    following_layers = []
+
+    for after_layer in layer_graph[layer].after:
+        follow(following_layers, layer_graph, layer_graph[after_layer])
+
+    # If you delete a part of the previous one, not just all of it then you cannot just say these n are to be deleted
+    # So for each layer you need to figure out what you are deleting from it, then for the next layer figure out what is deleted
+
+    deletions = {}
+    if isinstance(custom, list):
+        deletions[layer] = custom
+    elif isinstance(custom, int):
+        # The sorting magic happens here
+        # Find the first convolution
+        if sort:
+            if layer_graph[layer].type == 'bn':
+                breakpoint()
+            depthwise_skipped_following_layers = []
+            for after_layer in layer_graph[layer].after:
+                get_following_layers_skip_depthwise(depthwise_skipped_following_layers, layer_graph, layer_graph[after_layer])
+
+            first_conv = get_first_conv(layer_graph, depthwise_skipped_following_layers)
+
+            if first_conv is None:
+                breakpoint()
+
+            slices = get_relevent_slice(layer_graph, first_conv, layer)
+            if len(slices) == 0:
+                breakpoint()
+
+            importances = torch.zeros(layer_graph[layer].o).cuda()
+            for single_slice in slices:
+                importances += get_importances(layer_graph[first_conv].value.weight, single_slice)
+
+            importances = importances.tolist()
+            deletion_indices = pick_channels_with_lowest_importances(custom, importances)
+            if len(deletion_indices) != custom:
+                assert(False)
+            deletion_list = convert_to_deletions_list(deletion_indices)
+
+
+            #breakpoint()
+            deletions[layer] = deletion_list
+        else:
+            deletions[layer] = [(layer_graph[layer].o-custom, layer_graph[layer].o)]
+    else:
+        assert(False)
+
+    # If there is a plus operation, the other operand is a "mirror" of this
+    # SO copy whatever we do here to the other operand
+    for mirror in layer_graph[layer].mirror:
+        # Start a new custom deletion for the mirror
+        if reason != 'mirror':
```

Comment:
what does reason here even mean? what else can it be?

Reply:
This isn't clear still. I don't think I follow what a mirror or tie mean. Can you explain more and add an example?